solutions/cf-1713E.mdx

```diff
@@ -0,0 +1,162 @@
+---
+id: cf-1713E
+source: Codeforces Div 2
+title: Cross Swapping
+author: Aakash Gokhale
+---
+
+## Explanation
+
+[Official Analysis](https://codeforces.com/blog/entry/105419)
+
+Notice that $a[i][j]$ can only be swapped with $a[j][i]$ no matter which operations are performed.
+
+For all $j$ such that $i < j$ (we only need to consider each pair once and keeping $i < j$ will ensure this), we can determine whether we want to swap them or keep them the same.
+
+If $a[i][j] > a[j][i]$, we want to swap them. If $a[i][j] < a[j][i]$ we want to keep them the same. If they are equal it doesn't matter.
+
+However, it is possible that not all conditions can work together with each other and swapping something means we can't swap something else later on. Since we want the lexicographically smallest answer, the earlier indices matter more than later indices so we find optimal row swaps for earlier indices first.
```

Comment:
> not all conditions can work together with each other

what does this mean exactly?

Reply:
swapping/keeping the same $a[i][j]$ with $a[j][i]$ means we can't swap/keep the same $a[k][l]$ with $a[l][k]$ because $i = k$, $i = l$, $j = k$, $j = l$, or some combination.

Does this work as a replacement?

Reply:
However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same late on. For example it is impossible to swap $i$ and $j$, keep $j$ and $k$ the same, and keep $i$ and $k$ the same. We can first flip the $i$ith column to swap $i$ and $j$. We do nothing to $j$ or $k$ since we want to keep them the same. In order to keep $i$ and $k$ the same, we need to flip the $k$th column because we earlier swapped $i$ and two swaps will cancel out. However, swapping $k$ will mean that the second condition will fail and we have reached a scenario where we cannot satisfy all conditions.

Reply:
is this better?

Reply:
at this point i'd just give an example

Reply:
For example it is impossible to swap $i$ and $j$, keep $j$ and $k$ the same, and keep $i$ and $k$ the same.

Isn't this already an example?

Reply:
> swap $i$ and $j$

how can you swap $i$ and $j$? are these numbers? points?
instead of variables, just use concrete numbers, perhaps with the sample test case

Reply:
However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same late on. For example it is impossible to swap a[i][j] and a[j][i], keep a[j][k] and a[k][j] the same, and keep a[i][k] and a[k][i] the same. We can first flip the ith column/row which satisfies the first condition. We do nothing to the jth or kth column/row since we want to keep them the same which satisfies the second condition. In order to keep a[i][k] and a[k][i] the same, we need to flip the kth column/row and two swaps (i and k) will cancel out. However, swapping the kth column/row will cause the second condition to now fail and we have reached a scenario where we cannot satisfy all conditions.

Reply:
I was unclear in my explanation. I was supposed to reference a[i][j], a[j][k], and a[i][k]

Reply:
this is pretty hard to follow
i'd use the example input or make an example array & say "it's possible to swap 3 & 4, but then this would [something bad]"

Reply:
However, it is possible that not all conditions can work together with each other and swapping something/keeping something the same means we can't swap something else/keep something the same later on.
Consider to following input:
0 2 1
1 0 1
3 2 0
We want to first swap a[0][1] and a[1][0] because 1 is less than 2. We can do this by flipping the 0th row/col. This would give us the array:
0 1 3
2 0 1
1 2 0
Then we want to swap a[0][2] and a[2][0] because 1 is less than 3. To do this either we can either swap the 0th row/col or the 2nd row/col. However, swapping the 0th row/col would undo the efforts of our first operation so we will swap the 2nd row/col and get:
0 1 1
2 0 2
3 1 0
Now we want to swap a[1][2] and a[2][1] because 1 is less than 2 and we want the lexicographically minimum. However swapping the 1st row/col would mess up the swap we made in the first operation and swapping the 2nd row/col would mess up the swap we made in the second operation.
Thus, it is impossible to make all swaps optimally.

Is this better?

Reply:
much much better

Reply:
fixed